File: crates/verter_core/examples/ast.js
/**
 * Vue Template AST Generator
 *
 * This script reads .vue files from the source/ directory,
 * parses them using Vue's official compiler, and outputs
 * the AST to generated/{filename}.json
 *
 * Usage: node script.js
 *
 * Requirements: npm install @vue/compiler-sfc
 */

const fs = require("fs");
const path = require("path");
const { baseParse } = require("@vue/compiler-core");

const AST_DIR = path.join(__dirname, "ast");
const SOURCE_DIR = path.join(AST_DIR, "source");
const GENERATED_DIR = path.join(AST_DIR, "generated");

// Ensure generated directory exists
if (!fs.existsSync(GENERATED_DIR)) {
  fs.mkdirSync(GENERATED_DIR, { recursive: true });
}
/**
 * Recursively sort all object keys alphabetically
 * Uses a WeakSet to handle circular references
 */
function sortObjectKeys(obj, seen = new WeakSet()) {
  if (obj === null || typeof obj !== "object") {
    return obj;
  }

  // Handle circular references
  if (seen.has(obj)) {
    return "[Circular]";
  }
  seen.add(obj);

  if (Array.isArray(obj)) {
    return obj.map((item) => sortObjectKeys(item, seen));
  }

  const sorted = {};
  const keys = Object.keys(obj).sort();
  for (const key of keys) {
    sorted[key] = sortObjectKeys(obj[key], seen);
  }
  return sorted;
}

// Get all .vue files from source directory
const vueFiles = fs
  .readdirSync(SOURCE_DIR)
  .filter((file) => file.endsWith(".vue"));

if (vueFiles.length === 0) {
  console.log("No .vue files found in source/ directory");
  process.exit(0);
}

console.log(`Found ${vueFiles.length} .vue file(s) to process\n`);

for (const file of vueFiles) {
  const filePath = path.join(SOURCE_DIR, file);
  const baseName = path.basename(file, ".vue");
  const outputPath = path.join(GENERATED_DIR, `${baseName}.json`);

  console.log(`Processing: ${file}`);

  try {
    const source = fs.readFileSync(filePath, "utf-8");
    const ast = baseParse(source, {
      filename: file,
      sourceMap: false,
      templateParseOptions: {
        // whitespace is not implemented in verter core yet, but set it to preserve for future compatibility
        whitespace: "preserve",
      },
    });

    // if (errors.length > 0) {
    //   console.error(`  Errors parsing ${file}:`);
    //   errors.forEach((err) => console.error(`    - ${err.message}`));
    //   continue;
    // }

    const result = ast;

    // Sort keys alphabetically for easier comparison
    // const sortedResult = sortObjectKeys(result);
    const sortedResult = result;
    fs.writeFileSync(outputPath, JSON.stringify(sortedResult, null, 2));
    console.log(`  -> ${outputPath}`);
  } catch (err) {
    console.error(`  Error processing ${file}: ${err.message}`);
  }
}

console.log("\nDone!");

File: crates/verter_core/examples/ast.rs
//! AST Example
//!
//! This example reads .vue files from examples/ast/source/ directory,
//! parses them using verter's parser, and outputs the AST to
//! examples/ast/generated/{filename}.verter.json
//!
//! Run with: cargo run --example ast

use bumpalo::Bump;
use std::fs;
use std::path::Path;
use verter_core::syntax::plugin::SyntaxPluginOptions;
use vize_armature::parser as vize_parser;

fn main() {
    let example_dir = Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("examples")
        .join("ast");
    let source_dir = example_dir.join("source");
    let generated_dir = example_dir.join("generated");

    // Ensure generated directory exists
    fs::create_dir_all(&generated_dir).expect("Failed to create generated directory");

    // Get all .vue files from source directory
    let vue_files: Vec<_> = fs::read_dir(&source_dir)
        .expect("Failed to read source directory")
        .filter_map(|entry| entry.ok())
        .filter(|entry| {
            entry
                .path()
                .extension()
                .map(|ext| ext == "vue")
                .unwrap_or(false)
        })
        .collect();

    if vue_files.is_empty() {
        println!("No .vue files found in source/ directory");
        return;
    }

    println!("Found {} .vue file(s) to process\n", vue_files.len());

    let options = SyntaxPluginOptions::default();

    for entry in vue_files {
        let file_path = entry.path();
        let file_name = file_path.file_name().unwrap().to_string_lossy();
        let base_name = file_path.file_stem().unwrap().to_string_lossy();
        let output_path = generated_dir.join(format!("{}.verter.json", base_name));

        println!("Processing: {}", file_name);

        match fs::read_to_string(&file_path) {
            Ok(source) => {
                let allocator = oxc_allocator::Allocator::new();
                let root =
                    verter_core::builder::parser_syntax::parse(&source, &options, &allocator);

                // Serialize directly to pretty JSON
                let json_output =
                    serde_json::to_string_pretty(&root).expect("Failed to serialize JSON");
                fs::write(&output_path, &json_output).expect("Failed to write output file");
                println!("  -> {}", output_path.display());
            }
            Err(err) => {
                eprintln!("  Error reading file: {}", err);
            }
        }
    }

    println!("\nDone!");
    println!("\nTo compare with Vue's official parser, run:");
    println!("  cd examples/ast && npm install @vue/compiler-sfc && node script.js");
}
